lib/server.js

```diff
@@ -0,0 +1,131 @@
+// appended to the start of the given buffer
+function muxBuffer (code, buffer) {
+  return Buffer.concat([new Buffer(String.fromCharCode(code)), buffer]);
+}
+
+/** Makes an instance of docker exec running command, then links the provided
+    websocket to the stream provided by docker exec
+
+    @param tty: boolean representing whether or not terminal emulation is active
+  */
+function execWithSocket (container, socket, command, tty) {
+  var options = {
+    AttachStdout: true,
+    AttachStderr: true,
+    AttachStdin: true,
+    Tty: tty,
+    Detach: false,
+    Cmd: command,
+  };
+  var attachoptions = {
+    stdin: true,
+    stdout: true,
+    stderr: true,
+    stream: true,
+  };
+  return container.exec(options, (err, exec) => {
+    if (err) { //err proof
+      throw err;
+    }
+    exec.start(attachoptions, (err, execStream) => {
+      if (err) { //err proof
+        throw err;
+      }
+      var strout = through((data) => {
+        if (dbgstream) {
+          dbgstream.write('OUT: ' + data + '\n-\n');
+        }
+        socket.send(muxBuffer(1, data), {binary: true, mask: true});
+      });
+      var strerr = through((data) => {
+        if (dbgstream) {
+          dbgstream.write('ERR: ' + data + '\n-\n');
+        }
+        socket.send(muxBuffer(2, data), {binary: true, mask: true});
+      });
+      exec.modem.demuxStream(execStream, strout, strerr);
+
+      socket.on('message', (message) => {
+        if (dbgstream) {
+          dbgstream.write('IN: ' + message + '\n-\n');
+        }
+        execStream.write(message);
+      });
+
+      socket.on('disconnect', () => {
+        execStream.end('\x03\x04\r\nexit\r\n'); //should be how to ctrl+c ctrl+d, might be better way to kill
+      });
+
+      execStream.on('end', () => {
+        exec.inspect((err, data) => {
+          if (err) {
+            throw err;
+          }
+          var exitBuf = new Buffer(4);
+          exitBuf.writeUInt32LE(data.ExitCode);
+          socket.send(muxBuffer(100, exitBuf));
+        });
+      });
+
+      socket.send('ready'); //send ready message to let client know it is ready
+    });
+  });
+}
+
+export default class DockerExecWebsocketServer {
+  /* Creates Docker Exec instance on given container, running the first message given
+   * as a command.
+   * Options:
+   * port, required
+   * container, name or id of docker container, required
+   * path, path where the websocket is hosted
+   * tty, whether or not we expect VT100-style output
+   */
+  constructor (options) {
+    options = _.defaults({}, options, {path: '/'+slugid.v4(), tty: false});
+    var container = docker.getContainer(options.container);
+    this.port = options.port;
+    this.path = options.path;
+    var wsopts = {
+      port: this.port,
+      path: this.path,
+    };
+    assert(this.port, 'required port option missing');
+    assert(container, 'required container option missing');
+    this.server = new ws.Server(wsopts);
```

Comment:
Not yet important, but maybe worth considering, to try and steal the SSL configuration setup from here:
https://github.com/websockets/ws/blob/master/examples/ssl.js
So we already support SSL, then we can test it with a self-signed certificate for `localhost`, we just need to connect to that from a node script that has the option to validate certificate authority disabled... Or we can skip testing for starters.